File: apple-frameworks/MusicKit/Shims/MusicKitShims.cs
// Copyright (c) 2026 Justin Wojciechowski.
// Licensed under the MIT License.
//
// Hand-rolled partials that layer factory ctors onto the generated
// MusicCatalogSearchRequest, MusicLibrarySearchRequest, and
// MusicCatalogSearchSuggestionsRequest shells. The generator emits the
// storage + ISwiftObject plumbing but cannot synthesise initializers
// whose parameter is `[any P.Type]` — an array of existential metatypes
// (binding report skips them as "UnsupportedExistential").
//
// The companion Swift shim file (MusicKitShims.swift, staged into the
// per-framework wrapper) carries SBW_MusicKitShims_* @_cdecl trampolines
// that accept a stable UInt32 bitmask and decode it into the conformer
// `Type.self` value the Swift initializer wants. Bit ordering is
// append-only forever; unknown bits cause the shim to return non-zero
// and this partial to throw ArgumentException.
//
// Storage ownership matches the AttributedString shim pattern:
//
//   * C# allocates a heap slot sized by the request struct's value
//     witness table (via SwiftObjectHelper<T>.GetTypeMetadata().Size);
//   * the shim writes the initialized struct into the slot;
//   * SwiftSafeHandle takes sole ownership through the generated
//     private `T(SwiftHandle)` ctor and frees the slot on Dispose via
//     the type's destroy witness.

using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text;
using Swift.Runtime;

namespace MusicKit;

/// <summary>
/// Conformers of <c>MusicCatalogSearchable</c> usable as a
/// <see cref="MusicCatalogSearchRequest"/> result type filter. The bit
/// values form a stable, append-only ABI shared with the Swift shim —
/// future additions must take the next unassigned high bit.
/// </summary>
[Flags]
public enum MusicCatalogSearchTypes : uint
{
    None        = 0u,
    Album       = 1u << 0,
    Artist      = 1u << 1,
    Curator     = 1u << 2,
    MusicVideo  = 1u << 3,
    Playlist    = 1u << 4,
    RadioShow   = 1u << 5,
    RecordLabel = 1u << 6,
    Song        = 1u << 7,
    Station     = 1u << 8,
}

/// <summary>
/// Conformers of <c>MusicLibrarySearchable</c> usable as a
/// <see cref="MusicLibrarySearchRequest"/> result type filter. Stable,
/// append-only bit values mirroring the Swift shim's decoder.
/// </summary>
[Flags]
public enum MusicLibrarySearchTypes : uint
{
    None       = 0u,
    Album      = 1u << 0,
    Artist     = 1u << 1,
    MusicVideo = 1u << 2,
    Playlist   = 1u << 3,
    Song       = 1u << 4,
}

public partial class MusicCatalogSearchRequest
{
    private const uint CatalogSearchValidMask =
        (uint)(MusicCatalogSearchTypes.Album | MusicCatalogSearchTypes.Artist
            | MusicCatalogSearchTypes.Curator | MusicCatalogSearchTypes.MusicVideo
            | MusicCatalogSearchTypes.Playlist | MusicCatalogSearchTypes.RadioShow
            | MusicCatalogSearchTypes.RecordLabel | MusicCatalogSearchTypes.Song
            | MusicCatalogSearchTypes.Station);

    /// <summary>
    /// Constructs a <see cref="MusicCatalogSearchRequest"/> projection of
    /// the Swift <c>init(term: String, types: [any MusicCatalogSearchable.Type])</c>.
    /// The <paramref name="types"/> bitmask is decoded into the
    /// corresponding Swift metatype array inside the shim.
    /// </summary>
    /// <param name="term">Search term. A null reference is normalised
    /// to the empty string to match the empty-search behaviour of the
    /// Swift API.</param>
    /// <param name="types">Bitmask of MusicKit catalog item kinds to
    /// include in the result. A zero mask is legal and produces an
    /// unfiltered search (Apple's API also accepts <c>[]</c>).</param>
    /// <exception cref="ArgumentException">A bit outside the
    /// MusicCatalogSearchTypes enumeration is set.</exception>
    public static unsafe MusicCatalogSearchRequest Create(string term, MusicCatalogSearchTypes types)
    {
        uint stray = (uint)types & ~CatalogSearchValidMask;
        if (stray != 0)
        {
            throw new ArgumentException(
                $"MusicCatalogSearchTypes carries unknown bits: 0x{stray:X8}.",
                nameof(types));
        }
        var metadata = SwiftObjectHelper<MusicCatalogSearchRequest>.GetTypeMetadata();
        var heap = NativeMemory.Alloc((nuint)metadata.Size);
        bool initialized = false;
        try
        {
            var utf8 = string.IsNullOrEmpty(term)
                ? Array.Empty<byte>()
                : Encoding.UTF8.GetBytes(term);
            int status;
            fixed (byte* utf8Ptr = utf8)
            {
                status = MusicKitShimsNative.CatalogSearchRequestInit(
                    utf8Ptr, utf8.Length, (uint)types, heap);
            }
            if (status != 0)
            {
                throw new ArgumentException(
                    $"SBW_MusicKitShims_CatalogSearchRequest_Init returned status {status}.",
                    nameof(types));
            }
            initialized = true;
        }
        catch
        {
            if (initialized)
                metadata.ValueWitnessTable->Destroy(heap, metadata);
            NativeMemory.Free(heap);
            throw;
        }
        return new MusicCatalogSearchRequest(new SwiftHandle((IntPtr)heap));
    }
}

public partial class MusicLibrarySearchRequest
{
    private const uint LibrarySearchValidMask =
        (uint)(MusicLibrarySearchTypes.Album | MusicLibrarySearchTypes.Artist
            | MusicLibrarySearchTypes.MusicVideo | MusicLibrarySearchTypes.Playlist
            | MusicLibrarySearchTypes.Song);

    /// <summary>
    /// Constructs a <see cref="MusicLibrarySearchRequest"/> projection of
    /// the Swift <c>init(term: String, types: [any MusicLibrarySearchable.Type])</c>.
    /// </summary>
    /// <param name="term">Search term. Null is normalised to empty.</param>
    /// <param name="types">Bitmask of library item kinds to include.</param>
    /// <exception cref="ArgumentException">A bit outside the
    /// MusicLibrarySearchTypes enumeration is set.</exception>
    public static unsafe MusicLibrarySearchRequest Create(string term, MusicLibrarySearchTypes types)
    {
        uint stray = (uint)types & ~LibrarySearchValidMask;
        if (stray != 0)
        {
            throw new ArgumentException(
                $"MusicLibrarySearchTypes carries unknown bits: 0x{stray:X8}.",
                nameof(types));
        }
        var metadata = SwiftObjectHelper<MusicLibrarySearchRequest>.GetTypeMetadata();
        var heap = NativeMemory.Alloc((nuint)metadata.Size);
        bool initialized = false;
        try
        {
            var utf8 = string.IsNullOrEmpty(term)
                ? Array.Empty<byte>()
                : Encoding.UTF8.GetBytes(term);
            int status;
            fixed (byte* utf8Ptr = utf8)
            {
                status = MusicKitShimsNative.LibrarySearchRequestInit(
                    utf8Ptr, utf8.Length, (uint)types, heap);
            }
            if (status != 0)
            {
                throw new ArgumentException(
                    $"SBW_MusicKitShims_LibrarySearchRequest_Init returned status {status}.",
                    nameof(types));
            }
            initialized = true;
        }
        catch
        {
            if (initialized)
                metadata.ValueWitnessTable->Destroy(heap, metadata);
            NativeMemory.Free(heap);
            throw;
        }
        return new MusicLibrarySearchRequest(new SwiftHandle((IntPtr)heap));
    }
}

public partial class MusicCatalogSearchSuggestionsRequest
{
    private const uint CatalogSearchValidMask =
        (uint)(MusicCatalogSearchTypes.Album | MusicCatalogSearchTypes.Artist
            | MusicCatalogSearchTypes.Curator | MusicCatalogSearchTypes.MusicVideo
            | MusicCatalogSearchTypes.Playlist | MusicCatalogSearchTypes.RadioShow
            | MusicCatalogSearchTypes.RecordLabel | MusicCatalogSearchTypes.Song
            | MusicCatalogSearchTypes.Station);

    /// <summary>
    /// Constructs a <see cref="MusicCatalogSearchSuggestionsRequest"/>
    /// projection of the Swift
    /// <c>init(term: String, includingTopResultsOfTypes: [any MusicCatalogSearchable.Type])</c>.
    /// </summary>
    /// <param name="term">Search term. Null is normalised to empty.</param>
    /// <param name="includingTopResultsOfTypes">Bitmask of catalog
    /// item kinds to include as suggested top results. A zero mask is
    /// legal (Apple's API defaults to <c>[]</c>).</param>
    /// <exception cref="ArgumentException">A bit outside the
    /// MusicCatalogSearchTypes enumeration is set.</exception>
    public static unsafe MusicCatalogSearchSuggestionsRequest Create(
        string term, MusicCatalogSearchTypes includingTopResultsOfTypes)
    {
        uint stray = (uint)includingTopResultsOfTypes & ~CatalogSearchValidMask;
        if (stray != 0)
        {
            throw new ArgumentException(
                $"MusicCatalogSearchTypes carries unknown bits: 0x{stray:X8}.",
                nameof(includingTopResultsOfTypes));
        }
        var metadata = SwiftObjectHelper<MusicCatalogSearchSuggestionsRequest>.GetTypeMetadata();
        var heap = NativeMemory.Alloc((nuint)metadata.Size);
        bool initialized = false;
        try
        {
            var utf8 = string.IsNullOrEmpty(term)
                ? Array.Empty<byte>()
                : Encoding.UTF8.GetBytes(term);
            int status;
            fixed (byte* utf8Ptr = utf8)
            {
                status = MusicKitShimsNative.CatalogSearchSuggestionsRequestInit(
                    utf8Ptr, utf8.Length, (uint)includingTopResultsOfTypes, heap);
            }
            if (status != 0)
            {
                throw new ArgumentException(
                    $"SBW_MusicKitShims_CatalogSearchSuggestionsRequest_Init returned status {status}.",
                    nameof(includingTopResultsOfTypes));
            }
            initialized = true;
        }
        catch
        {
            if (initialized)
                metadata.ValueWitnessTable->Destroy(heap, metadata);
            NativeMemory.Free(heap);
            throw;
        }
        return new MusicCatalogSearchSuggestionsRequest(new SwiftHandle((IntPtr)heap));
    }
}

internal static unsafe partial class MusicKitShimsNative
{
    [LibraryImport("MusicKitSwiftBindings", EntryPoint = "SBW_MusicKitShims_CatalogSearchRequest_Init")]
    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
    public static partial int CatalogSearchRequestInit(byte* utf8Ptr, int utf8Len, uint mask, void* outBuffer);

    [LibraryImport("MusicKitSwiftBindings", EntryPoint = "SBW_MusicKitShims_LibrarySearchRequest_Init")]
    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
    public static partial int LibrarySearchRequestInit(byte* utf8Ptr, int utf8Len, uint mask, void* outBuffer);

    [LibraryImport("MusicKitSwiftBindings", EntryPoint = "SBW_MusicKitShims_CatalogSearchSuggestionsRequest_Init")]
    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
    public static partial int CatalogSearchSuggestionsRequestInit(byte* utf8Ptr, int utf8Len, uint mask, void* outBuffer);
}

File: apple-frameworks/MusicKit/Shims/MusicKitShims.swift
// Copyright (c) 2026 Justin Wojciechowski.
// Licensed under the MIT License.
//
// @_cdecl trampolines that project MusicKit's `[any P.Type]`-typed search
// request constructors into a calling-convention the generator can speak.
//
// The Swift initializers in question are:
//
//   * MusicCatalogSearchRequest.init(term: String, types: [any MusicCatalogSearchable.Type])
//   * MusicLibrarySearchRequest.init(term: String, types: [any MusicLibrarySearchable.Type])
//   * MusicCatalogSearchSuggestionsRequest.init(term: String,
//                                               includingTopResultsOfTypes types: [any MusicCatalogSearchable.Type])
//
// The last is the sibling search-suggestions ctor; it shares the same
// `[any MusicCatalogSearchable.Type]` shape and reuses the same
// bit-to-type mapping.
//
// The `[any P.Type]` parameter is a heterogeneous array of existential
// metatypes that the binding generator cannot project (binding report
// emits "UnsupportedExistential" for these inits). The shims accept a
// stable UInt32 bitmask whose bits index into a fixed, append-only set
// of conformer metatypes — one mapping table per protocol. The C# face
// is a `[Flags]` enum per protocol so call sites read naturally:
//
//     MusicCatalogSearchRequest.Create(term, MusicCatalogSearchTypes.Song | .Album)
//
// Encoding is append-only forever: new conformers in a future SDK get
// the next high bit. Unknown bits cause the shim to return a non-zero
// status so the C# wrapper can throw ArgumentException rather than
// silently dropping the bit.
//
// Storage ownership matches the existing SBW_AttributedString_Init
// pattern: the caller (C#) allocates a heap slot sized by the request
// struct's value witness table, the shim writes the initialized struct
// into the slot, and a SwiftSafeHandle takes sole ownership. Mutations
// on the C# side see refcount 1 and apply in place.
//
// Availability gating: the shim is compiled against the binding's
// wrapper deployment target (currently iOS 15.0 / macOS 12.0 /
// macCatalyst 15.0 / tvOS 15.0). Each @_cdecl entry point is annotated
// with the @available band of the highest-versioned API it references.
// Within decoders, conformer appends introduced in later point releases
// are wrapped in `if #available` so the catalog-search decoder remains
// callable at the iOS 15.0 baseline.

import MusicKit
import Foundation

// MARK: - MusicCatalogSearchable bit mapping (9 conformers)
//
// Bits assigned in alphabetical order of the conforming type name. This
// order is the public ABI contract — any reshuffle is a breaking change
// for already-shipped binaries. New conformers from future MusicKit
// releases must take the next unassigned high bit.
//
//   bit 0 — Album        (iOS 15.0 / macOS 12.0 / tvOS 15.0)
//   bit 1 — Artist       (iOS 15.0 / macOS 12.0 / tvOS 15.0)
//   bit 2 — Curator      (iOS 15.4 / macOS 12.3 / tvOS 15.4)
//   bit 3 — MusicVideo   (iOS 15.4 / macOS 12.3 / tvOS 15.4)
//   bit 4 — Playlist     (iOS 15.0 / macOS 12.0 / tvOS 15.0)
//   bit 5 — RadioShow    (iOS 15.4 / macOS 12.3 / tvOS 15.4)
//   bit 6 — RecordLabel  (iOS 15.0 / macOS 12.0 / tvOS 15.0)
//   bit 7 — Song         (iOS 15.0 / macOS 12.0 / tvOS 15.0)
//   bit 8 — Station      (iOS 15.0 / macOS 12.0 / tvOS 15.0)

private let SBW_MUSICKIT_CATALOG_SEARCH_VALID_MASK: UInt32 = 0x1FF

@available(iOS 15.0, macOS 12.0, macCatalyst 15.0, tvOS 15.0, *)
@inline(__always)
private func decodeCatalogSearchableTypes(_ mask: UInt32) -> [any MusicCatalogSearchable.Type] {
    var out: [any MusicCatalogSearchable.Type] = []
    if mask & (1 << 0) != 0 { out.append(Album.self) }
    if mask & (1 << 1) != 0 { out.append(Artist.self) }
    if mask & (1 << 2) != 0 {
        if #available(iOS 15.4, macOS 12.3, macCatalyst 15.4, tvOS 15.4, *) {
            out.append(Curator.self)
        }
    }
    if mask & (1 << 3) != 0 {
        if #available(iOS 15.4, macOS 12.3, macCatalyst 15.4, tvOS 15.4, *) {
            out.append(MusicVideo.self)
        }
    }
    if mask & (1 << 4) != 0 { out.append(Playlist.self) }
    if mask & (1 << 5) != 0 {
        if #available(iOS 15.4, macOS 12.3, macCatalyst 15.4, tvOS 15.4, *) {
            out.append(RadioShow.self)
        }
    }
    if mask & (1 << 6) != 0 { out.append(RecordLabel.self) }
    if mask & (1 << 7) != 0 { out.append(Song.self) }
    if mask & (1 << 8) != 0 { out.append(Station.self) }
    return out
}

// MARK: - MusicLibrarySearchable bit mapping (5 conformers)
//
// All five conformances are gated by the MusicLibrarySearchable
// protocol itself (iOS 16.0 / macOS 14.0 / macCatalyst 17.0 / tvOS 16.0).
//
//   bit 0 — Album
//   bit 1 — Artist
//   bit 2 — MusicVideo
//   bit 3 — Playlist
//   bit 4 — Song

private let SBW_MUSICKIT_LIBRARY_SEARCH_VALID_MASK: UInt32 = 0x1F

@available(iOS 16.0, macOS 14.0, macCatalyst 17.0, tvOS 16.0, *)
@inline(__always)
private func decodeLibrarySearchableTypes(_ mask: UInt32) -> [any MusicLibrarySearchable.Type] {
    var out: [any MusicLibrarySearchable.Type] = []
    if mask & (1 << 0) != 0 { out.append(Album.self) }
    if mask & (1 << 1) != 0 { out.append(Artist.self) }
    if mask & (1 << 2) != 0 { out.append(MusicVideo.self) }
    if mask & (1 << 3) != 0 { out.append(Playlist.self) }
    if mask & (1 << 4) != 0 { out.append(Song.self) }
    return out
}

@inline(__always)
private func decodeUtf8Term(_ utf8Ptr: UnsafePointer<UInt8>?, _ utf8Len: Int) -> String {
    if let utf8Ptr = utf8Ptr, utf8Len > 0 {
        return String(decoding: UnsafeBufferPointer(start: utf8Ptr, count: utf8Len), as: UTF8.self)
    }
    return ""
}

// MARK: - @_cdecl trampolines
//
// Each trampoline returns Int32:
//   0  — success, outBuffer holds the initialized struct (caller owns)
//   -1 — `mask` carries bits outside the valid range (programming error;
//        C# wrapper throws ArgumentException)
//
// outBuffer is NOT written when the return value is non-zero, so the
// caller must NOT run the destroy witness on the slot in that case.

/// Initialize a `MusicCatalogSearchRequest(term:types:)` into the
/// caller-provided heap slot. `mask` is a UInt32 bitmask over
/// SBW_MUSICKIT_CATALOG_SEARCH_VALID_MASK (see top of file for the
/// bit-to-type mapping).
@available(iOS 15.0, macOS 12.0, macCatalyst 15.0, tvOS 15.0, *)
@_cdecl("SBW_MusicKitShims_CatalogSearchRequest_Init")
public func SBW_MusicKitShims_CatalogSearchRequest_Init(
    _ utf8Ptr: UnsafePointer<UInt8>?,
    _ utf8Len: Int,
    _ mask: UInt32,
    _ outBuffer: UnsafeMutableRawPointer
) -> Int32 {
    if mask & ~SBW_MUSICKIT_CATALOG_SEARCH_VALID_MASK != 0 { return -1 }
    let term = decodeUtf8Term(utf8Ptr, utf8Len)
    let types = decodeCatalogSearchableTypes(mask)
    let req = MusicCatalogSearchRequest(term: term, types: types)
    outBuffer.assumingMemoryBound(to: MusicCatalogSearchRequest.self).initialize(to: req)
    return 0
}

/// Initialize a `MusicLibrarySearchRequest(term:types:)` into the
/// caller-provided heap slot. See SBW_MUSICKIT_LIBRARY_SEARCH_VALID_MASK
/// for the bit-to-type mapping. Available on iOS 16 / macOS 14 /
/// macCatalyst 17 / tvOS 16+; older deployment targets cannot reach
/// this symbol even though it links into the wrapper.
@available(iOS 16.0, macOS 14.0, macCatalyst 17.0, tvOS 16.0, *)
@_cdecl("SBW_MusicKitShims_LibrarySearchRequest_Init")
public func SBW_MusicKitShims_LibrarySearchRequest_Init(
    _ utf8Ptr: UnsafePointer<UInt8>?,
    _ utf8Len: Int,
    _ mask: UInt32,
    _ outBuffer: UnsafeMutableRawPointer
) -> Int32 {
    if mask & ~SBW_MUSICKIT_LIBRARY_SEARCH_VALID_MASK != 0 { return -1 }
    let term = decodeUtf8Term(utf8Ptr, utf8Len)
    let types = decodeLibrarySearchableTypes(mask)
    let req = MusicLibrarySearchRequest(term: term, types: types)
    outBuffer.assumingMemoryBound(to: MusicLibrarySearchRequest.self).initialize(to: req)
    return 0
}

/// Initialize a
/// `MusicCatalogSearchSuggestionsRequest(term:includingTopResultsOfTypes:)`
/// into the caller-provided heap slot. The `includingTopResultsOfTypes`
/// array is decoded from the same MusicCatalogSearchable bit mapping —
/// a zero mask is legal and corresponds to the Swift API's documented
/// default of `[]`. Available on iOS 16 / macOS 13 / macCatalyst 16 /
/// tvOS 16+; older deployment targets cannot link this entry point even
/// if the binding is loaded.
@available(iOS 16.0, macOS 13.0, macCatalyst 16.0, tvOS 16.0, *)
@_cdecl("SBW_MusicKitShims_CatalogSearchSuggestionsRequest_Init")
public func SBW_MusicKitShims_CatalogSearchSuggestionsRequest_Init(
    _ utf8Ptr: UnsafePointer<UInt8>?,
    _ utf8Len: Int,
    _ topResultsMask: UInt32,
    _ outBuffer: UnsafeMutableRawPointer
) -> Int32 {
    if topResultsMask & ~SBW_MUSICKIT_CATALOG_SEARCH_VALID_MASK != 0 { return -1 }
    let term = decodeUtf8Term(utf8Ptr, utf8Len)
    let types = decodeCatalogSearchableTypes(topResultsMask)
    let req = MusicCatalogSearchSuggestionsRequest(term: term, includingTopResultsOfTypes: types)
    outBuffer.assumingMemoryBound(to: MusicCatalogSearchSuggestionsRequest.self).initialize(to: req)
    return 0
}
